Added references in PeerAuth, AuthPolicy, ServiceEntry and Sidecar.

hhovsepy opened the pull request:
RFE https://github.com/kiali/kiali/issues/2447

AuthorizationPolicy objects, added references to Workloads, Services, ServiceEntries and VirtualServices.
ServiceEntry objects, added references to AuthorizationPolicies and DestinationRules.
Sidecar objects, references to Workloads, Services and ServiceEntries.
PeerAuth objects, references to DestinationRules and Workloads.

UI changes are required for testing: https://github.com/kiali/kiali-ui/pull/2299

Xunzhuo commented on business/references/auth_policy_references.go:
```diff
@@ -0,0 +1,113 @@
+			}
+			if len(rule.To) > 0 {
+				for _, t := range rule.To {
+					if t == nil {
+						continue
+					}
+					if t.Operation == nil {
+						continue
+					}
+					if len(t.Operation.Hosts) == 0 {
+						continue
+					}
```
Can these lines be merged like this?
``` golang
if t == nil || t.Operation == nil || len(t.Operation.Hosts) == 0 {
     continue
}
```

leandroberetta commented:
Hello @hhovsepy,  I started testing this PR with the new side panel UI.

I created an "allow nothing" auth policy and I saw that there aren't workload references, I was hoping to see all the workloads from the namespace listed, what do you think about this?. In the UI I will need to add a way to collapse references if the list is big.

Also, I tested the following auth policy:

```yaml
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: allow-nothing
  namespace: bookinfo
spec:
  selector:
    matchLabels:
      app: reviews
```
 
I was expecting to see all workloads of the review app (v1, v2 and v3) but I only see the first:

![sp_ap_ref_allow_nothing_reviews](https://user-images.githubusercontent.com/1286393/155386052-cbe0d078-1535-4def-8e4a-02861a72aad6.png)

In the workloads list I can see that the all 3 are covered:

![sp_ap_ref_w_list](https://user-images.githubusercontent.com/1286393/155386231-32dbb568-127d-4ba7-9aa6-b0b46e2d4262.png)

lucasponce reviewed (dismissed):
The code looks good.
I haven't tested all the cases, as this work will need the UI, but it's ok to consolidate.

These are the tests I've performed:
- Some basic regression testing using "travels" and "fraud-detection" apps; I've performed the tutorial and all features worked fine.
- For this particular feature I've tested some endpoints manually to check the references:

-- Destination Rule:
![image](https://user-images.githubusercontent.com/1662329/156152419-7736f88d-2a61-47da-8f89-9e69713ed791.png)

-- Sidecar:
![image](https://user-images.githubusercontent.com/1662329/156152473-ede9707d-c1e2-4ea5-aa2f-f127b6fd8ea7.png)

-- Authorization Policy:
![image](https://user-images.githubusercontent.com/1662329/156152548-ca98cbf1-52d3-493f-b960-ba8b8de181d6.png)
![image](https://user-images.githubusercontent.com/1662329/156152587-cbf1d4c7-bcba-4511-945a-1a14083bb206.png)
![image](https://user-images.githubusercontent.com/1662329/156152645-11061120-ab35-4662-976f-25c59a389ef3.png)

The title of the PR should be changed as it updates more references.

I'll put this PR in the QE queue, for visibility, they can take a look and decide if the current testing is enough to merge or we should perform other test on it.

hhovsepy commented:
@lucasponce I have added more references in this PR, especially PeerAuth references to DR and Workloads.
Fixed some missing references and tests from other objects.
Ready for a final review.

lucasponce commented:
@hhovsepy I see a regression in the config icons of the IstioConfig list.

There are some warnings in the overview:
![image](https://user-images.githubusercontent.com/1662329/157203154-1ac18b42-a4ea-4501-902f-fd892d2a1a33.png)

But in the list there is a N/A:
![image](https://user-images.githubusercontent.com/1662329/157203131-331405ca-fb7f-4080-b53b-60b706a96b0a.png)

Probably it's not related to the references, but as we are doing more deep testing looking for regressions in every PR, it would be worth to take a look.

hhovsepy commented:
Rebased with master, it seems problem with config status is gone.

lucasponce commented:
I'm still seeing the N/A in the services
![image](https://user-images.githubusercontent.com/1662329/157237019-8e8f7418-3a67-419c-8c82-ed1ee5c46978.png)

lucasponce commented:
Found another issue to investigate around the services + service entry.

I'm using the "electronic-shop" demo which is prepared for SE scenarios:

![image](https://user-images.githubusercontent.com/1662329/157239954-c9c6a977-7dd7-4e94-8f15-286c1675298e.png)

Then invoke the validations endpoint around the SE:

```
GET http://localhost:8000/api/namespaces/electronic-shop/services/*.msn.com?validate=true
```
The validations response is populated with data not linked with the SE:

```
{
  "service": {
    "name": "*.msn.com",
    "createdAt": "",
    "resourceVersion": "",
    "namespace": {
      "name": "electronic-shop",
      "labels": null,
      "annotations": null
    },
    "labels": null,
    "selectors": null,
    "type": "External",
    "ip": "",
    "ports": [
      {
        "name": "http-port",
        "protocol": "HTTP",
        "istioProtocol": "",
        "port": 80,
        "tlsMode": ""
      },
      {
        "name": "https",
        "protocol": "HTTPS",
        "istioProtocol": "",
        "port": 443,
        "tlsMode": ""
      }
    ],
    "externalName": "",
    "healthAnnotations": {},
    "additionalDetails": null
  },
  "istioSidecar": true,
  "endpoints": null,
  "virtualServices": [],
  "destinationRules": [],
  "serviceEntries": [
    {
      "kind": "ServiceEntry",
      "apiVersion": "networking.istio.io/v1alpha3",
      "metadata": {
        "name": "zone-msn-com",
        "namespace": "electronic-shop",
        "resourceVersion": "24835",
        "creationTimestamp": "2022-03-08T12:35:04Z",
        "annotations": {
          "kubectl.kubernetes.io/last-applied-configuration": "{\"apiVersion\":\"networking.istio.io/v1alpha3\",\"kind\":\"ServiceEntry\",\"metadata\":{\"annotations\":{},\"name\":\"zone-msn-com\",\"namespace\":\"electronic-shop\"},\"spec\":{\"hosts\":[\"*.msn.com\"],\"location\":\"MESH_EXTERNAL\",\"ports\":[{\"name\":\"http-port\",\"number\":80,\"protocol\":\"HTTP\",\"targetPort\":443},{\"name\":\"https\",\"number\":443,\"protocol\":\"HTTPS\"}],\"resolution\":\"NONE\"}}\n"
        }
      },
      "spec": {
        "hosts": [
          "*.msn.com"
        ],
        "ports": [
          {
            "number": 80,
            "protocol": "HTTP",
            "name": "http-port",
            "targetPort": 443
          },
          {
            "number": 443,
            "protocol": "HTTPS",
            "name": "https"
          }
        ]
      },
      "status": {}
    }
  ],
  "istioPermissions": {
    "create": true,
    "update": true,
    "delete": true
  },
  "workloads": [],
  "health": {
    "requests": {
      "inbound": {
        "http": {
          "301": 0.5318522167487685
        }
      },
      "outbound": {},
      "healthAnnotations": {}
    }
  },
  "validations": {
    "destinationrule": {
      "control.travel-control": {
        "name": "control",
        "objectType": "destinationrule",
        "valid": true,
        "checks": [],
        "references": null
      }
    },
    "serviceentry": {
      "zone-msn-com.electronic-shop": {
        "name": "zone-msn-com",
        "objectType": "serviceentry",
        "valid": true,
        "checks": [],
        "references": null
      }
    },
    "sidecar": {
      "cars-v1.travel-agency": {
        "name": "cars-v1",
        "objectType": "sidecar",
        "valid": true,
        "checks": [
          {
            "code": "KIA0004",
            "message": "No matching workload found for the selector in this namespace",
            "severity": "warning",
            "path": "spec/workloadSelector/labels"
          }
        ],
        "references": null
      },
      "discounts-v1.travel-agency": {
        "name": "discounts-v1",
        "objectType": "sidecar",
        "valid": true,
        "checks": [
          {
            "code": "KIA0004",
            "message": "No matching workload found for the selector in this namespace",
            "severity": "warning",
            "path": "spec/workloadSelector/labels"
          }
        ],
        "references": null
      },
      "flights-v1.travel-agency": {
        "name": "flights-v1",
        "objectType": "sidecar",
        "valid": true,
        "checks": [
          {
            "code": "KIA0004",
            "message": "No matching workload found for the selector in this namespace",
            "severity": "warning",
            "path": "spec/workloadSelector/labels"
          }
        ],
        "references": null
      },
      "hotels-v1.travel-agency": {
        "name": "hotels-v1",
        "objectType": "sidecar",
        "valid": true,
        "checks": [
          {
            "code": "KIA0004",
            "message": "No matching workload found for the selector in this namespace",
            "severity": "warning",
            "path": "spec/workloadSelector/labels"
          }
        ],
        "references": null
      },
      "insurances-v1.travel-agency": {
        "name": "insurances-v1",
        "objectType": "sidecar",
        "valid": true,
        "checks": [
          {
            "code": "KIA0004",
            "message": "No matching workload found for the selector in this namespace",
            "severity": "warning",
            "path": "spec/workloadSelector/labels"
          }
        ],
        "references": null
      },
      "mysqldb-v1.travel-agency": {
        "name": "mysqldb-v1",
        "objectType": "sidecar",
        "valid": true,
        "checks": [
          {
            "code": "KIA0004",
            "message": "No matching workload found for the selector in this namespace",
            "severity": "warning",
            "path": "spec/workloadSelector/labels"
          }
        ],
        "references": null
      },
      "mysqldb-v2.travel-agency": {
        "name": "mysqldb-v2",
        "objectType": "sidecar",
        "valid": true,
        "checks": [
          {
            "code": "KIA0004",
            "message": "No matching workload found for the selector in this namespace",
            "severity": "warning",
            "path": "spec/workloadSelector/labels"
          }
        ],
        "references": null
      },
      "mysqldb-v3.travel-agency": {
        "name": "mysqldb-v3",
        "objectType": "sidecar",
        "valid": true,
        "checks": [
          {
            "code": "KIA0004",
            "message": "No matching workload found for the selector in this namespace",
            "severity": "warning",
            "path": "spec/workloadSelector/labels"
          }
        ],
        "references": null
      },
      "travels-v1.travel-agency": {
        "name": "travels-v1",
        "objectType": "sidecar",
        "valid": true,
        "checks": [
          {
            "code": "KIA0004",
            "message": "No matching workload found for the selector in this namespace",
            "severity": "warning",
            "path": "spec/workloadSelector/labels"
          }
        ],
        "references": null
      },
      "travels-v2.travel-agency": {
        "name": "travels-v2",
        "objectType": "sidecar",
        "valid": true,
        "checks": [
          {
            "code": "KIA0004",
            "message": "No matching workload found for the selector in this namespace",
            "severity": "warning",
            "path": "spec/workloadSelector/labels"
          }
        ],
        "references": null
      },
      "travels-v3.travel-agency": {
        "name": "travels-v3",
        "objectType": "sidecar",
        "valid": true,
        "checks": [
          {
            "code": "KIA0004",
            "message": "No matching workload found for the selector in this namespace",
            "severity": "warning",
            "path": "spec/workloadSelector/labels"
          }
        ],
        "references": null
      }
    },
    "virtualservice": {
      "bookinfo.default": {
        "name": "bookinfo",
        "objectType": "virtualservice",
        "valid": true,
        "checks": [],
        "references": null
      },
      "control.travel-control": {
        "name": "control",
        "objectType": "virtualservice",
        "valid": true,
        "checks": [],
        "references": null
      }
    },
    "workload": {
      "bookstore-v1.electronic-shop": {
        "name": "bookstore-v1",
        "objectType": "workload",
        "valid": true,
        "checks": [],
        "references": null
      },
      "games-v1.electronic-shop": {
        "name": "games-v1",
        "objectType": "workload",
        "valid": true,
        "checks": [],
        "references": null
      },
      "games-v2.electronic-shop": {
        "name": "games-v2",
        "objectType": "workload",
        "valid": true,
        "checks": [],
        "references": null
      },
      "mobile.electronic-shop": {
        "name": "mobile",
        "objectType": "workload",
        "valid": true,
        "checks": [],
        "references": null
      },
      "music-v1.electronic-shop": {
        "name": "music-v1",
        "objectType": "workload",
        "valid": true,
        "checks": [],
        "references": null
      },
      "shop-v1.electronic-shop": {
        "name": "shop-v1",
        "objectType": "workload",
        "valid": true,
        "checks": [],
        "references": null
      },
      "web.electronic-shop": {
        "name": "web",
        "objectType": "workload",
        "valid": true,
        "checks": [],
        "references": null
      }
    }
  },
  "namespaceMTLS": {
    "status": "MTLS_NOT_ENABLED"
  }
}
```

I suspect it's due the SE has a wildcard as part of the name and that may skip some host filter.

lucasponce reviewed (commented):
The code looks good to me.

Tested PA:
![image](https://user-images.githubusercontent.com/1662329/157241687-806d2e17-15ff-4f6f-9027-71a33639934a.png)

Tested AP:
![image](https://user-images.githubusercontent.com/1662329/157241748-ec190316-49ab-4782-b7b2-7e9e6079661c.png)

Tested Sidecars:
![image](https://user-images.githubusercontent.com/1662329/157241843-47b49409-6538-4ce7-b47d-d8947f394f35.png)

lucasponce commented:
In the case of the ServiceEntries we miss a bidirectional reference, I don't see it a blocker as it's a special case, but this is what we see from Kiali:

![image](https://user-images.githubusercontent.com/1662329/157242343-72f893b0-f7dd-4866-8cc2-6e09076d03c7.png)

A Service in the Istio Registry that is defined from an Istio ServiceEntry.

In the Service details we have it as part of the Istio definition:

![image](https://user-images.githubusercontent.com/1662329/157242415-ef98ca9e-77a2-4e3a-94d6-738ddd523b59.png)

But in the ServiceEntry details:

```
GET http://localhost:8000/api/namespaces/electronic-shop/istio/serviceentries/zone-msn-com?validate=true
```

There are no references to the Istio Registry:

![image](https://user-images.githubusercontent.com/1662329/157242576-3b492d99-7d80-468b-9263-d92d24f8990d.png)

I don't see it as a blocker, I'm open to consolidating this logic to avoid the risk of conflicts. The UI part is not yet ready, but I comment in these two cases to investigate if those could lead to any other problem that I haven't seen during this review.

hhovsepy commented:
Added the bidirectional references to Services from ServiceEntry Istio object details.

lucasponce commented:
Tested the last commit.

Now, the ServiceEntry has a reference to the service.
![image](https://user-images.githubusercontent.com/1662329/157414263-c9dfef34-95cc-4eca-9b70-5dca15d34314.png)

lucasponce reviewed (approved):
I'm +1 to consolidate this work.

Note that this has not yet the UI ready https://github.com/kiali/kiali-ui/pull/2299 but it can be merged.

When the UI work is close to finishing, we can review the end-to-end cases.

jshaughn reviewed (approved):
I ran through the Travel tutorial as a regression test.  None found.  Approving.

Non-blockers and/or outside of this PR: 

I did get one unexplained, "could not fetch namespaces" alert, I think it was after I had confirmed the addition of the travels service routing rules.  I could not reproduce it.

The validations counter for the namespace in overview can indicate several checked validations (I expected N/A, but looks like cross-namespace counter is shown) but when navigating to the config page for the namespace there are no config objects.  It is a little confusing and something we need to address outside of this PR.

I also noticed that there may be an issue with edge label values surviving when crossing zoom thresholds.
